### PR TITLE
Extract cluster IP and POD IP with cluster management APIs

### PR DESCRIPTION
**Pull request by seswarrajan**
Signed-off-by: Eswar Rajan Subramanian <eswar@accuknox.com>

### CODE REVIEW

File: src/libs/mysqlHandler.go [hunk 1]
```diff
@@ -79,12 +79,12 @@ func GetNetworkPoliciesFromMySQL(cfg types.ConfigDB, cluster, namespace, status,
 	}
 
 	results, err = DBHandle.Query(query+whereClause, args...)
-	defer results.Close()
 
 	if err != nil {
 		log.Error().Msg(err.Error())
 		return nil, err
 	}
+	defer results.Close()
```
Comment (nyrahul) [inline src/libs/mysqlHandler.go:88]: This change seems to be incorrect. The `Close()` should be done only if the `Query()` is successful. The `results.Close()` should not be done if the `err != nil`.

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: src/libs/dbHandler.go [hunk 2]
```diff
@@ -1,11 +1,57 @@
...
+// WaitForDB
+func WaitForDB(db *sql.DB) {
+	for {
+		err := db.Ping()
+		if err != nil {
+			time.Sleep(time.Second * 1)
+			log.Error().Msgf("DB Ping() failed. Will retry. err=%s", err.Error())
+		} else {
+			break
+		}
```
Comment (nyrahul) [inline src/libs/dbHandler.go]:
```suggestion
		err := db.Ping()
		if err == nil {
			break
		}
		time.Sleep(time.Second * 1)
		log.Error().Msgf("DB Ping() failed. Will retry. err=%s", err.Error())
```

File: src/libs/dbHandler.go [hunk 3]
```diff
@@ -1,11 +1,57 @@
...
+func NewMockDB() (*sql.DB, sqlmock.Sqlmock) {
+	db, mock, err := sqlmock.New()
+	if err != nil {
+		log.Error().Msgf("an error '%s' was not expected when opening a stub database connection", err)
+	}
+
+	MockSql = mock
+	MockDB = db
+
+	return db, mock
```
Comment (nyrahul) [inline src/libs/dbHandler.go]: Is there a point in returning the db pointers as well as setting the global values? This is not a good practice.

File: src/libs/dbHandler.go [hunk 4]
```diff
@@ -1,11 +1,57 @@
...
+			time.Sleep(time.Second * 1)
+			log.Error().Msgf("DB Ping() failed. Will retry. err=%s", err.Error())
+		} else {
+			break
+		}
+	}
+}
+
+func NewMockDB() (*sql.DB, sqlmock.Sqlmock) {
+	db, mock, err := sqlmock.New()
```
Comment (nyrahul) [inline src/libs/dbHandler.go]: Why to add this function is `dbHandler.go` when it is used only from `_test` file?

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: src/libs/dbHandler.go (hunk 4 above)
Comment (seswarrajan) [reply src/libs/dbHandler.go]: @nyrahul This function existed in both DB files. Hence maintained in the same location.

File: src/libs/dbHandler.go (hunk 3 above)
Comment (seswarrajan) [reply src/libs/dbHandler.go]: This is as part of existing code. Hence cleaning up. Will handle this.

File: src/libs/mysqlHandler.go (hunk 1 above)
Comment (seswarrajan) [reply src/libs/mysqlHandler.go:88]: Had the same confusion, but moved this handle as per existing db functions. Will handle this.

File: src/libs/dbHandler.go (hunk 4 above)
Comment (nyrahul) [reply src/libs/dbHandler.go]: But according to this commit, this is a newly added function?

File: src/libs/mysqlHandler.go (hunk 1 above)
Comment (seswarrajan) [reply src/libs/mysqlHandler.go:88]: Screenshot from 2022-07-25 16-35-04

Comment (seswarrajan) [reply src/libs/mysqlHandler.go:88]: Suggestion is to check error before deferring results.Close() or stmt.Close()

File: src/libs/dbHandler.go (hunk 4 above)
Comment (seswarrajan) [reply src/libs/dbHandler.go]: Yes @nyrahul
This was earler handled in sqlitehandler and mysqlhandler.go.

Hence thought of moving to a common position dbhandler.go. 
But currently moved to dbhandler_test.go. Thanks for pointing it out.

**Review by wazir-ahmed — CHANGES_REQUESTED** [review]
@seswarrajan 
The changes looks good to me. But I think we missed a scenario with the DB connection handler. We only connect once during `InitDB()`. What will happen if the connection is disconnected (due to some network failure)? We have to periodically `Ping()` and check if the connection is active, if not, we have to reconnect.

We also have to protect the `DBHandler` pointer with some kind of lock since it could get updated & accessed from two goroutines at the same time (The reconnect logic updating the pointer and sql functions accessing the pointer).

**Comment by wazir-ahmed** [comment]
@nyrahul @seswarrajan This PR is becoming bigger day by day. It has also become a PR about many things. Please don't add any more commit on this PR. Please merge this PR if this is tested.

**Comment by seswarrajan** [comment]
> 

Yes @wazir-ahmed 
I just wanted to make sure not to miss any changes, so making this a draft PR. Will try to split and raise PR for relevant topics.

**Pull request closed without merging**